### DISCUSSION TITLE
Flat eslint config: can't understand a way how it is intended to used, it's design is wrong and it is not documented as expected

### DISCUSSION THREAD

**Discussion by andrew-aladjev** [typescript-eslint/typescript-eslint#8881] [Technical Discussions]
### Body

I am trying to migrate to Eslint `9.0.0`. I want to prepare a flat eslint configuration based on the old `.eslintrc.json`. I've read the flat config documentation and found that the new config consists of objects that will be applied to each source file. Let's go deeper:

```js
import typescriptEslint from 'typescript-eslint';
console.log(typescriptEslint.configs.all);
```

```
[
  {
    languageOptions: { parser: ..., sourceType: 'module' },
    plugins: { '@typescript-eslint': ... }
  },
  {
    files: [ '**/*.ts', '**/*.tsx', '**/*.mts', '**/*.cts' ],
    rules: {
      'constructor-super': 'off',
      ...
    }
  },
  {
    rules: { ... },
  }
]
```

We can see that default configuration consists of objects:

1. The first object includes just language options, parser, parser options, and plugin.
2. The second object includes rule exceptions that should be applied to special files only and should not be applied to any other files.
3. The third object includes rules that should be applied to all files.

I've read several pages of documentation and found a recommended "way" to prepare the configuration:

```js
console.log(typescriptEslint.config({
  files: ['src/**/*'],
  languageOptions: { sourceType: 'other', parserOptions: { sourceType: 'other' } },
  plugins: { 'my-plugin': 'my-plugin' },
  extends: typescriptEslint.configs.all,
}));
```

```
[
  {
    languageOptions: { parser: [Object], sourceType: 'module' },
    plugins: { '@typescript-eslint': [Object] },
    files: [ 'src/**/*' ]
  },
  {
    files: [ 'src/**/*' ],
    rules: {
      'constructor-super': 'off',
      ...
    }
  },
  {
    rules: { ... },
    files: [ 'src/**/*' ]
  },
  {
    files: [ 'src/**/*' ],
    languageOptions: { sourceType: 'other', parserOptions: [Object] },
    plugins: { 'my-plugin': 'my-plugin' }
  }
]
```

1. First object has just limited the set of source files where parser should be applied. I am not able to change language or parser options for typescript-eslint parser. I am not able to add plugin toward typescript-eslint plugin.
2. Second rule is broken completely: we have wider range of files where exceptions should be applied.
3. Third rules is valid, we have just limited the set of files for all rules.
4. 4-th object has no sense: it applies my-plugin together with some options for all files.

I am pretty sure that `typescript-eslint` should not recommend `typescriptEslint.config` function in current form.

### Before you submit your discussion, please confirm the following. If any of these required steps are not taken, we may not be able to review your RFC. Help us to help you!

- [X] I have [read the discussions guidelines](https://typescript-eslint.io/contributing/discussions) and per those guidelines, this fits a discussion category, _not_ a help request or standard issue.
- [X] I have [searched for related discussions](https://github.com/typescript-eslint/typescript-eslint/discussions) and [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues) and found none that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

**Comment by bradzacher** [comment]
As per our [discussion guidelines](https://typescript-eslint.io/contributing/discussions) we do not use discussions for support requests.

Please visit [one of these alternatives](https://typescript-eslint.io/contributing/issues#questions-and-support-requests) to get support.
